### PR TITLE
Make Layer a Node, add LayerGroup class

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
Thanks to help from @brisvag, we're going to try pushing on layer groups again.  #2261 has gotten hard to reconcile with main, so this PR pulls out most of the good stuff in there into a new PR.  This will be just a step along the way, and this PR _should_ be an "invisible" change.  It does swap out `Layerlist` for `LayerGroup`, and `QtLayerList` for `QtLayerTree`... but it disables actually adding a nested layer group or list to `viewer.layers` unless you specifically opt in with `ALLOW_LAYERGROUPS=1`.  Those two changes are also nicely isolated one-line changes.

### What remains to be done (in this PR):
- [ ] layers appear reversed again (with higher indices lower down in the list, need to apply the reverse proxy model)
- [ ] there's some duplicated logic on the Qt side between the flat QtLayerList, and the nested one.  need to review the inheritance to reduce code duplication
- [ ] fix remaining test failures
- [ ] review all methods on `LayerList` and make sure that LayerGroup implements them (layergroups are both a layerlist and a layer)
- [ ] decide on naming:
   - should it be called `LayerGroup`?  `LayerTree`?  `NestedLayerList`?
   - since layergroup is now itself a layer.  i think it belongs in `napari/layers` ... thoughts?  export in `napari.layers`?  or only in `napari.layers.layergroup`?

### What remains to be done (in future PRs):
- may a vispy LayerGroup node.  With this PR, the view has no idea how to render a LayerGroup, but @brisvag has played with some proof of principle stuff here, and I think we have a relatively clear vision of how we can also implement the composite pattern that this brings to the layerlist model in the view, with a proper scenegraph pattern
- there is still a bug in the selection model with nested layers... need to debug

this shouldn't go in for 0.4.14 ... but probably right after, so devs can start using it just to make sure that the standard (non-nested) use case is fine.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/components/layerlist.py
Comment (sofroniewn) [inline napari/components/layerlist.py:36]: We should remove those other deprecated methods - i believe here https://github.com/napari/napari/blob/6a3e11aa717e7928a0a5a3c7693577729a466ef1/napari/layers/base/base.py#L317-L330
and then drop this

Comment (sofroniewn) [reply napari/components/layerlist.py:36]: ah just realized this method already existed and is just moving

**Review by sofroniewn — COMMENTED** [review]
Code changes are looking good. I just flagged one thing about deprecations, but that's actually not specific to this PR. I can pull this down and play with it to check all functionality is preserved, but agreed would be good to get this in soon!!

File: napari/components/layerlist.py
Comment (tlambert03) [reply napari/components/layerlist.py:36]: yeah, we can't actually remove this yet, because we still use them internally here:
https://github.com/napari/napari/blob/6a3e11aa717e7928a0a5a3c7693577729a466ef1/napari/layers/base/base.py#L1679-L1689

i.e. it's still waiting on changes to highlighting in the layer model

File: napari/utils/events/containers/_nested_list.py
Comment (Czaki) [inline napari/utils/events/containers/_nested_list.py]:
```suggestion
        for i, item in enumerate(self[:stop]):
            yield root + (i,) if root else i
            if isinstance(item, NestableEventedList):
                yield from item._iter_indices(root=root + (i,))
```

**Review by Czaki — CHANGES_REQUESTED** [review]

File: napari/_qt/containers/_base_item_model.py
Comment (andy-sweet) [inline napari/_qt/containers/_base_item_model.py]: Remove?

File: napari/_qt/containers/_base_item_model.py
Comment (andy-sweet) [inline napari/_qt/containers/_base_item_model.py]: Consider removing this implementation comment because the code around it already describes what's going on here. Instead add an implementation comment to summarize what the two code blocks above are doing, because that's not clear to me (even with some previous experience with `QModelIndex`).

File: napari/layers/layergroup/layergroup.py
Comment (andy-sweet) [inline napari/layers/layergroup/layergroup.py:22]: `LayerGroup` sub-typing `Layer` makes me uncomfortable (for reasons I'm struggling to coherently describe), but my guess it's almost required to make progress here - is that right?

I think it's because I think of `Layer.data` as a VIP in napari, but `LayerGroup` is a `Layer` that I assume we expect will always have no data. I think this kind of mirrors the separation of Node/Group as Leaf/Composite, in that we only expect leaf nodes in the layer group tree to have data.

**Review by andy-sweet — COMMENTED** [review]
It's hard for me to review this as the PR touches quite a lot of existing files, most of which I'm not too familiar with. I initially left a couple of semi-useless line comments, so didn't add any more.

From my perspective, the most important things to focus on here are how does this change the public API of napari. In particular, what does `viewer.layers` return when I execute it in the napari console and is the API of that thing structurally different than it was before. I'm OK if `type(viewer.layers)` evaluates to `LayerGroup` instead of `LayerList` now, less OK if we expanded the public API (I think we have), and not OK at all if we broken the public API. After spending some time review this, it's not clear where we are right now, so a summary of those changes would be really useful.

The other thing I would find really useful is a summary of what the plan is moving forward. That definitely shouldn't block this PR from getting merged, but going from a flat layer list to a nested one feels like a sea change that should have some coherent proposal around it. If I understand the comments around this PR, I think the nesting/grouping here is pretty closely connected to the scene graph in vispy (and its relatives) in that the parent-child relationship, but I think there is also a concept of layer groups in Photoshop (which I'm not familiar with), so I'm not sure how much overlap there is here (e.g. does opacity multiply through to child layers like transforms would?).

**Comment by andy-sweet** [comment]
I also pulled this down to check if anything broke without `ALLOW_LAYERGROUPS=1`. The only thing I found was that right-clicking a layer raises an Exception, and if I click on the View Traceback button, I get the segfault. That might be a known bug, but definitely needs to be fixed before this gets merged.

<details><summary>Traceback + segfault details: </summary><p>

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/projects/napari/napari/_qt/containers/_layer_delegate.py in editorEvent(self=<napari._qt.containers._layer_delegate.LayerDelegate object>, event=<PyQt5.QtGui.QMouseEvent object>, model=<napari._qt.containers.qt_layer_tree.QtLayerTreeModel object>, option=<PyQt5.QtWidgets.QStyleOptionViewItem object>, index=<PyQt5.QtCore.QModelIndex object>)
    150             and event.button() == Qt.RightButton
    151         ):
--> 152             self.show_context_menu(
        self.show_context_menu = <bound method LayerDelegate.show_context_menu of <napari._qt.containers._layer_delegate.LayerDelegate object at 0x7fd245d02940>>
        index = <PyQt5.QtCore.QModelIndex object at 0x7fd22a7aa4a0>
        model = <napari._qt.containers.qt_layer_tree.QtLayerTreeModel object at 0x7fd245d028b0>
        event.globalPos = <built-in method globalPos of QMouseEvent object at 0x7fd240805940>
        option.widget = <napari._qt.containers.qt_layer_tree.QtLayerTreeView object at 0x7fd245d02820>
    153                 index, model, event.globalPos(), option.widget
    154             )

~/projects/napari/napari/_qt/containers/_layer_delegate.py in show_context_menu(self=<napari._qt.containers._layer_delegate.LayerDelegate object>, index=<PyQt5.QtCore.QModelIndex object>, model=<napari._qt.containers.qt_layer_tree.QtLayerTreeModel object>, pos=PyQt5.QtCore.QPoint(150, 504), parent=<napari._qt.containers.qt_layer_tree.QtLayerTreeView object>)
    181             self._context_menu = QtActionContextMenu(_LAYER_ACTIONS)
    182 
--> 183         layer_list: LayerList = model.sourceModel()._root
        layer_list = undefined
        global LayerList = undefined
        model.sourceModel._root = undefined
    184         self._context_menu.update_from_context(get_context(layer_list))
    185         action = self._context_menu.exec_(pos)

AttributeError: 'QtLayerTreeModel' object has no attribute 'sourceModel'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~/projects/napari/napari/_qt/containers/_layer_delegate.py in editorEvent(self=<napari._qt.containers._layer_delegate.LayerDelegate object>, event=<PyQt5.QtGui.QMouseEvent object>, model=<napari._qt.containers.qt_layer_tree.QtLayerTreeModel object>, option=<PyQt5.QtWidgets.QStyleOptionViewItem object>, index=<PyQt5.QtCore.QModelIndex object>)
    150             and event.button() == Qt.RightButton
    151         ):
--> 152             self.show_context_menu(
        self.show_context_menu = <bound method LayerDelegate.show_context_menu of <napari._qt.containers._layer_delegate.LayerDelegate object at 0x7fd245d02940>>
        index = <PyQt5.QtCore.QModelIndex object at 0x7fd249fdfa50>
        model = <napari._qt.containers.qt_layer_tree.QtLayerTreeModel object at 0x7fd245d028b0>
        event.globalPos = <built-in method globalPos of QMouseEvent object at 0x7fd240805940>
        option.widget = <napari._qt.containers.qt_layer_tree.QtLayerTreeView object at 0x7fd245d02820>
    153                 index, model, event.globalPos(), option.widget
    154             )

~/projects/napari/napari/_qt/containers/_layer_delegate.py in show_context_menu(self=<napari._qt.containers._layer_delegate.LayerDelegate object>, index=<PyQt5.QtCore.QModelIndex object>, model=<napari._qt.containers.qt_layer_tree.QtLayerTreeModel object>, pos=PyQt5.QtCore.QPoint(148, 503), parent=<napari._qt.containers.qt_layer_tree.QtLayerTreeView object>)
    181             self._context_menu = QtActionContextMenu(_LAYER_ACTIONS)
    182 
--> 183         layer_list: LayerList = model.sourceModel()._root
        layer_list = undefined
        global LayerList = undefined
        model.sourceModel._root = undefined
    184         self._context_menu.update_from_context(get_context(layer_list))
    185         action = self._context_menu.exec_(pos)

AttributeError: 'QtLayerTreeModel' object has no attribute 'sourceModel'
/Users/asweet/software/miniconda3/envs/napari-dev/bin/pythonw: line 3: 26162 Segmentation fault: 11  /Users/asweet/software/miniconda3/envs/napari-dev/python.app/Contents/MacOS/python "$@"
```
</p>

File: napari/layers/layergroup/layergroup.py
Comment (tlambert03) [reply napari/layers/layergroup/layergroup.py:22]: yep, this is the composite pattern that @jni proposed a while back and we all decided we liked.  This is the thing that makes it possible for vispy, for example, to _only_ interact with the layergroup (which does the work of deferring to it's components).  It's pairs perfectly with the scenegraph idea, where the top level scene need only pay attention to top level nodes, and needn't recurse all the way down. 

I feel your hesitation.  so feel free to keep contemplating that and poking at it.  So far, that particular part hasn't been problematic, and is pretty well covered by tests here.  but definitely open to thoughts

File: napari/_qt/containers/_base_item_model.py
Comment (tlambert03) [reply napari/_qt/containers/_base_item_model.py]: that whole block is probably going away, but will revise the comment there too

**Comment by tlambert03** [comment]
thanks @andy-sweet, I know this is a complex one... I appreciate you having a look

> From my perspective, the most important things to focus on here are how does this change the public API of napari. In particular, what does `viewer.layers` return when I execute it in the napari console and is the API of that thing structurally different than it was before

from an API perspective, this shouldn't change any of the existing behaviors of `viewer.layers`.   It still behaves as a mutable sequence, with all of the same methods as before (this much is also in the tests).   Technically, it goes from being a `LayerList` to a `LayerGroup` (the latter being a superset of the former). This just _adds_ the ability to add `LayerGroup` objects (or simply lists) to `viewer.layers`.  So, whereas this would have been an error before:

```
viewer.layers.append([Image(), Points()])
```

it would be allowed now.  The python model for the tree structure here is just a list of layers-or-lists (recursively).

> The other thing I would find really useful is a summary of what the plan is moving forward.  ... so I'm not sure how much overlap there is here (e.g. does opacity multiply through to child layers like transforms would?).

Sorry, there's been quite a bit of previous discussion on this, and I agree it would be great to go through the previous issues and summarize stuff.  (see https://github.com/napari/napari/issues/970, https://github.com/napari/napari/pull/1089 as a start)

> I also pulled this down to check if anything broke without ALLOW_LAYERGROUPS=1 ...The only thing I found was that right-clicking a layer raises an Exception, 

yes, this is partially related to the trying to fix the first bullet points in the original post "layers appear reversed again".   it's in an intermediate state while i try to make the reverse proxy model work again

**Comment by brisvag** [comment]
> > The other thing I would find really useful is a summary of what the plan is moving forward.  ... so I'm not sure how much overlap there is here (e.g. does opacity multiply through to child layers like transforms would?).
> 
> Sorry, there's been quite a bit of previous discussion on this, and I agree it would be great to go through the previous issues and summarize stuff. (see #970, #1089 as a start)

I can give a short summary of what I think we need to do once this is merged:

1. create `VispyLayerGroup`, which is a simple empty node in the vispy scenegraph. I already implemeted [some form of this elsewhere](https://github.com/brisvag/napari/blob/layergroups/napari/_vispy/layers/group.py), and everything worked fine. We get the expected tree-like behaviour for visibility, opacity and transforms "for free", because they act at the `node` level in vispy.
2. update the way `QtViewer` communicates with vispy. Currently, the communication is one-way only (events > vispy updates), and compartimentalized (each `VispyLayer` is only aware of itself and the canvas). We need to update this so that `VispyLayer` can be aware (even just temporarily, through events), about other `VispyLayer`s; this way, we can update the `VispyLayer.node.parent` to the correct `VispyLayerGroup`, and not just the main canvas.
3. make sure things like `transform` behave in a sensible way (e.g: if you drag and drop a layer in a new group, recompute the transform so that it leads to the same place, but relative to the new node? Or just use the raw transform and now the data is in a different place?
4. Implement other linkiing stuff and decide how to handle it: what happens if I have two image layers in a group? Do I have a single `contrast limits` slider that sets both? Are they always "locked in" on the same value, or do they stay separate if the layegroup one is unset? And what should happen when we have a vectors and a points? How do we distinguish overlapping parameters like `edge_color`?
4. get this stuff to populate the gui

I think we can get point 1 and 2 relatively easily once this is merged, and thus get the basic behaviour that many of us want, which is the simple ability to group related things and hide them when needed.

**Comment by andy-sweet** [comment]
> from an API perspective, this shouldn't change any of the existing behaviors of `viewer.layers`. It still behaves as a mutable sequence, with all of the same methods as before (this much is also in the tests). Technically, it goes from being a `LayerList` to a `LayerGroup` (the latter being a superset of the former). This just _adds_ the ability to add `LayerGroup` objects (or simply lists) to `viewer.layers`. So, whereas this would have been an error before:

Then this PR looks good to me on this point.

**Comment by sofroniewn** [comment]
@tlambert03 I'm wondering if after the 0.4.16 release comes out you want to pair on this and try to get it merged? Is the checklist in the description up to date?

**Comment by andy-sweet** [comment]
Closing in favor of #5132

**Pull request closed without merging**
